flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/compact/AggregationMergeFunction.java

```diff
@@ -0,0 +1,150 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.store.file.mergetree.compact;
+
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A {@link MergeFunction} where key is primary key (unique) and value is the partial record, update
+ * non-null fields on merge.
+ */
+@SuppressWarnings("checkstyle:RegexpSingleline")
+public class AggregationMergeFunction implements MergeFunction {
+
+    private static final long serialVersionUID = 1L;
+
+    private final RowData.FieldGetter[] getters;
+
+    private final RowType rowType;
+    private final ArrayList<AggregateFunction<?>> aggregateFunctions;
+    private final boolean[] isPrimaryKey;
+    private final RowType primaryKeyType;
+    private transient GenericRowData row;
+
+    private final Set<String> aggregateColumnNames;
+
+    public AggregationMergeFunction(
+            RowType primaryKeyType, RowType rowType, Set<String> aggregateColumnNames) {
+        this.primaryKeyType = primaryKeyType;
+        this.rowType = rowType;
+        this.aggregateColumnNames = aggregateColumnNames;
+
+        List<LogicalType> fieldTypes = rowType.getChildren();
+        this.getters = new RowData.FieldGetter[fieldTypes.size()];
+        for (int i = 0; i < fieldTypes.size(); i++) {
+            getters[i] = RowData.createFieldGetter(fieldTypes.get(i), i);
+        }
+
+        this.isPrimaryKey = new boolean[this.getters.length];
+        Arrays.fill(isPrimaryKey, false);
+        List<String> rowNames = rowType.getFieldNames();
+        for (String primaryKeyName : primaryKeyType.getFieldNames()) {
+            isPrimaryKey[rowNames.indexOf(primaryKeyName)] = true;
+        }
+
+        this.aggregateFunctions = new ArrayList<>(rowType.getFieldCount());
+        for (int i = 0; i < rowType.getFieldCount(); i++) {
+            AggregateFunction<?> f = null;
+            if (aggregateColumnNames.contains(rowNames.get(i))) {
+                f = choiceRightAggregateFunction(rowType.getTypeAt(i).getDefaultConversion());
+            }
+            aggregateFunctions.add(f);
+        }
+    }
+
+    private AggregateFunction<?> choiceRightAggregateFunction(Class<?> c) {
+        AggregateFunction<?> f = null;
+        if (Double.class.equals(c)) {
+            f = new DoubleAggregateFunction();
+        } else if (Long.class.equals(c)) {
+            f = new LongAggregateFunction();
+        } else if (Integer.class.equals(c)) {
+            f = new IntegerAggregateFunction();
+        } else if (Float.class.equals(c)) {
+            f = new FloatAggregateFunction();
+        }
+        return f;
+    }
```

Comment:
A better implementation is to have a factory or util class which reads in a aggregate function name and its type and gives out the corresponding function.